Adds unit test for user.py

Kausam opened the pull request:
Adds unit test for user module
Fix #203 .

Changes proposed in this pull request:
- unit tests for functions nick_change_graph, top_keywords_for_nick, keywords, and extended_stop_words of lib/analysis/user.py

## Checks
- [x] Single commit and [No merge commits](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Make sure you have added required documentation
- [ ] If you have refactored a code, make sure that you have compared the outputs pre and post refactoring. Add both pre and post refaoring screenshots below if the output is visual

## Any images?

## Notify reviewers
@rohangoel96

prasadtalasila commented:
@bhuvangupta008  your review of the suggested changes would be good.

kaushik-rohit commented:
@Kausam Some suggestions from quick look of PR.
1)  Please squash both commits into 1.
2) Don't use titles like adds unit-test for 3/5 function in user.py. Rather you can write "introduce unitests for user module" as title and list down the functions for which unit-test was included as bullets in the body.
3) One of the test is failing because you are calling os.remove on something that doesnot exist. Make sure that is correct.

Commit:
add unit test to user module

Kausam commented:
@kaushik-rohit Thanks for the review. I have dropped the old, obsolete commit and modified the commit message.
But the os.remove() is there to delete a temporary file which is created during the testing of keywords_clusters function. Moreover, the test is not throwing any error when I am testing it locally. I am not sure why it does so on travis. Do you have any suggestions?

kaushik-rohit commented:
@Kausam The path you have provided to os.remove is not correct. You are calling os.remove("data/user_test/temp_output_keywords_clusters.txt")
 but "data/user_test/temp_output_keywords_clusters.txt" is not a valid path.
You need to provide the absolute path. The way it is done in other files. os.path.join(current_directory, path relative to current_directory)

bhuvangupta008 commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,138 @@
+        mock_build_graphs.side_effect = build_graphs_util 
+        nick_change_graph_list = user.nick_change_graph(self.log_data, True)
+        expected_nick_change_graph_list = util.load_from_disk(current_directory + "/data/user_test/nick_change_graph_list")
+        for i in range(len(nick_change_graph_list)):
```
Conisder using [xrange](https://stackoverflow.com/questions/94935/what-is-the-difference-between-range-and-xrange-functions-in-python-2-x).

Kausam commented:
@prasadtalasila, I have incorporated the changes suggested by @bhuvangupta008 and @kaushik-rohit . Is there anything else pending in this pull request?

kaushik-rohit commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
+import filecmp
 
 current_directory = os.path.dirname(os.path.realpath(__file__))
+state0 = 0
```
please change this variable name to probably something like build_graphs_util_call_counter or something short. state0 seems vague.

kaushik-rohit commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
 
 current_directory = os.path.dirname(os.path.realpath(__file__))
+state0 = 0
+state1 = 0
```
Ditto as above.

kaushik-rohit commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
+        assert expected_user_words_dict == user_words_dict
+        assert expected_nicks_for_stop_words == nicks_for_stop_words
+        assert expected_sorted_keywords_for_channels == sorted_keywords_for_channels
+
```
it is better to use self.assertEqual()

kaushik-rohit commented:
@Kausam Please see the suggested changes. Other changes looks fine to me.

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
+import sys
+import filecmp
 
 current_directory = os.path.dirname(os.path.realpath(__file__))
```
@kausam
Why do we have global variables in this module? Anything the unit tests need can be put into the test setup function. Of course, remember to have a proper tear down method as well.

prasadtalasila replied:
These variables can become attributes of UserTest class.

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
 current_directory = os.path.dirname(os.path.realpath(__file__))
+state0 = 0
+state1 = 0
+def build_graphs_util(nick_sender, nick_receiver, time, year, month, day, day_graph, aggr_graph):
```
Considering that the `build_graphs_util()` function is used only by the functions of UserTest, it is better to move this function to inside the UserTest class and make it private with two underscores prefix. Please see the notation for private functions [here](http://radek.io/2011/07/21/private-protected-and-public-in-python/) and [here](https://stackoverflow.com/questions/7609508/is-it-a-convention-to-prefix-private-classes-with-underscores).

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
+        sys.stdout = captured_output
+        user.keywords_clusters(self.log_data, self.nicks_hash, self.nick_same_list, current_directory + "/data/user_test/","temp_output_keywords_clusters")
+        sys.stdout = sys.__stdin__
+        f = open(current_directory + "/data/user_test/stdout_captured_output_keywords_clusters.txt", "r")    
```
These three lines can be replaced with util.load_from_disk() function.

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,135 @@
+        f.close()
+        output = captured_output.getvalue() 
+        assert expected_captured_output == output 
+        assert filecmp.cmp(current_directory + "/data/user_test/output_keywords_clusters.txt", current_directory + "/data/user_test/temp_output_keywords_clusters.txt")
```
@kaushik-rohit and @bhuvangupta008 This pattern of matching seems fairly readable. Instead of reading from file into an object and comparing with the object generated, we can also put the object value into a file using util.save_to_disk() function and then on both the files we can apply filecmp.cmp() function.

Obviously, when we are performing file comparison this method makes more sense. When we are doing object comparison, doing object equality makes more sense. We can acknowledge the use of this file comparison pattern and put it in coding style document.

@bhuvangupta008 you can update the coding style document.

prasadtalasila reviewed (changes requested):
@Kausam please see the requested changes.

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,146 @@
+        output = captured_output.getvalue()
+
+        self.assertEqual(expected_captured_output, output)
+        self.assertTrue(filecmp.cmp(current_directory + "/data/user_test/output_keywords_clusters.txt", current_directory + "/data/user_test/temp_output_keywords_clusters.txt"))
```
@Kausam Based on our preferences, we are going towards reading the file into memory and asserting on the equality of the contents.

prasadtalasila commented on test/unit-test/test_lib/test_analysis/test_user.py:
```diff
@@ -1,34 +1,146 @@
+        self.build_graphs_util_call_counter = None
+
+
+    def __build_graphs_util(nick_sender, nick_receiver, time, year, month, day, day_graph, aggr_graph):
```
Please change the signature of the function to return an object and test based on that object.
